### PR TITLE
INT: `Add format string argument` intention

### PR DESCRIPTION
**Pull request by ortem**
Merge after #3753 

fmt_string_arg

### CODE REVIEW

**Comment by sirgl** [comment]
Would be even more great with completion :)

**Comment by ortem** [comment]
@sirgl What exactly do you mean? Code completion is supported inside the popup window.

**Comment by ortem** [comment]
bors retry

File: src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/RsAddFmtStringArgumentPopup.kt [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                super.removeKeyListener(l)
+                editor?.contentComponent?.removeKeyListener(l)
+            }
+        }
+        editorTextField.setFontInheritedFromLAF(false)
+        editorTextField.font = EditorUtil.getEditorFont()
+        return editorTextField
+    }
+
+    private fun showBalloon(editor: Editor,
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/RsAddFmtStringArgumentPopup.kt]: Formatting

File: src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/RsAddFmtStringArgumentPopup.kt [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            .setHideOnAction(false)
+            .setHideOnKeyOutside(false)
+            .setFillColor(UIUtil.getPanelBackground())
+            .setBorderInsets(JBUI.insets(3))
+            .createBalloon()
+        Disposer.register(parent, balloon)
+
+        val keyListener = object : KeyAdapter() {
+            override fun keyPressed(e: KeyEvent) {
+                if (e.keyCode == KeyEvent.VK_ENTER) {
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/RsAddFmtStringArgumentPopup.kt]: `when`

File: src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/RsAddFmtStringArgumentPopup.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                    onComplete.run()
+                } else if (e.keyCode == KeyEvent.VK_ESCAPE) {
+                    balloon.hide()
+                }
+            }
+        }
+        editorTextField.addKeyListener(keyListener)
+
+        val fontMetrics = editorTextField.getFontMetrics(editorTextField.font)
+        val minimalWidth = fontMetrics.stringWidth("1234")
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/RsAddFmtStringArgumentPopup.kt]: "1234"?

File: src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/AddFmtStringArgumentIntention.kt [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+        val result = placeholderRegex.findAll(oldStringUntilCaret)
+        val placeholderNumber = result.count()
+
+        val codeFragment = RsExpressionCodeFragment(project, codeFragmentText, macroCallExpr)
+
+        if (ApplicationManager.getApplication().isUnitTestMode) {
+            addFmtStringArgument(project, editor, ctx, codeFragment, caretOffsetInLiteral, placeholderNumber)
+        } else {
+            RsAddFmtStringArgumentPopup.show(editor, project, codeFragment, Runnable {
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/AddFmtStringArgumentIntention.kt]: You can replace `Runnable` with `() -> Unit` in function signature.

File: src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/AddFmtStringArgumentIntention.kt [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+        project.runWriteCommandAction {
+            ctx.macroCall.replace(newMacroCall) as RsMacroCall
+            editor.caretModel.moveToOffset(editor.caretModel.offset + newPlaceholder.length)
+        }
+    }
+
+    companion object {
+        private val formatMacros = setOf("format", "write", "writeln", "print", "println", "eprint", "eprintln", "format_args")
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/AddFmtStringArgumentIntention.kt]:
```suggestion
private val FORMAT_MACROS: Set<String> =
    hashSetOf("format", "write", "writeln", "print", "println", "eprint", "eprintln", "format_args")
```

File: src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/AddFmtStringArgumentIntention.kt [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        var codeFragmentText: String = "" // for tests
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/intentions/addFmtStringArgument/AddFmtStringArgumentIntention.kt]: `@VisibleForTesting`

**Comment by ice1000** [comment]
This feature bad bad

**Pull request merged**